### PR TITLE
[primTorch] added ref for nn.functional.glu

### PR DESCRIPTION
**Pull request by fdrocha**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82819
* __->__ #82214
* #83524

### CODE REVIEW

File: torch/_refs/nn/functional/__init__.py [hunk 1]
```diff
@@ -559,3 +560,16 @@ def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
     # It may be better to use clamp here, but we use hardtanh to replicate
     # the behavior of the existing implementation
     return refs.nn.functional.hardtanh(a, 0, 6)
+
+
+@register_decomposition(torch.ops.aten.glu)
+@out_wrapper()
+def glu(a: TensorLikeType, dim: int = -1) -> TensorLikeType:
+    dim = utils.canonicalize_dims(a.ndim, dim)
+    if a.shape[dim] % 2:
```
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:569]: use `check` here

File: torch/_refs/nn/functional/__init__.py [hunk 2]
```diff
@@ -559,3 +560,16 @@ def relu6(a: TensorLikeType, inplace: bool = False) -> TensorLikeType:
     # It may be better to use clamp here, but we use hardtanh to replicate
     # the behavior of the existing implementation
     return refs.nn.functional.hardtanh(a, 0, 6)
+
+
+@register_decomposition(torch.ops.aten.glu)
+@out_wrapper()
```
Comment (ngimel) [inline torch/_refs/nn/functional/__init__.py:566]: you need type promotion wrapper here

File: torch/testing/_internal/common_methods_invocations.py [hunk 3]
```diff
@@ -21078,6 +21078,12 @@ def __init__(
                          dtypes=(torch.float32,), device_type='cpu'),
         ),
     ),
+    PythonRefInfo(
+        "_refs.nn.functional.glu",
+        torch_opinfo_name="nn.functional.glu",
+        supports_nvfuser=False,
```
Comment (ngimel) [inline torch/testing/_internal/common_methods_invocations.py:21084]: cc @IvanYashchuk, this seems like an easy pointwise operation, I'm surprised nvfuser doesn't support it

Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py:21084]: `torch.tensor_split` is not supported yet.
The error is `Node slice_in_dim_default with target prims.slice_in_dim.default does not support nvfuser`.
`prims.slice_in_dim` is used in the `torch._refs.tensor_split` implementation.

File: test/test_decomp.py [hunk 4]
```diff
@@ -159,6 +159,8 @@ def op_assert_ref(test_case, op, test_dtype, i, orig, decomp, ref, args, kwargs)
         (torch.float16, torch.ops.aten.native_batch_norm.default): 1e-5,
         (torch.bfloat16, torch.ops.aten.linalg_vector_norm.default): 1e-6,
         (torch.float16, torch.ops.aten.linalg_vector_norm.default): 1e-6,
+        (torch.float16, torch.ops.aten.glu.default): 1e-2,
+        (torch.bfloat16, torch.ops.aten.glu.default): 1e-2,
```
Comment (fdrocha) [inline test/test_decomp.py:163]: One thing I wanted to check explicitly is this. The tests don't pass unless I make the precision this low. It's only for {b,}float16 so maybe it's okay - what do you think @ngimel ?

Comment (Lezcano) [reply test/test_decomp.py:163]: fwiw, this comes from the fact that the sigmoid is computed internally without casting yhe intermediate results to float32 as we do in TI

Comment (ngimel) [reply test/test_decomp.py:163]: Oh you are right @Lezcano, we should fix that in the eager implementation

Comment (Lezcano) [reply test/test_decomp.py:163]: I think eager is fine here, we should fix the ref. 
In the ref, we should first cast the inputs to float32, perform the computations in float32, and then cast the result back to whatever lower precision type we had.

Note that this behaviour is shared by all the TI implementations in eager, so this could very well be implemented as a decorator.

Comment (IvanYashchuk) [reply test/test_decomp.py:163]: Just add type promotion wrapper here as was already suggested https://github.com/pytorch/pytorch/pull/82214#discussion_r939167379 and tests will pass without tolerance overrides.
```py
@elementwise_type_promotion_wrapper(
    type_promoting_args=("a",),
    type_promotion_kind=ELEMENTWISE_TYPE_PROMOTION_KIND.DEFAULT,
)
```

Comment (Lezcano) [reply test/test_decomp.py:163]: ag, that's it, thank you :)

Comment (fdrocha) [reply test/test_decomp.py:163]: Yes, thanks!

Comment (ngimel) [reply test/test_decomp.py:163]: Eager doesn't do type promotion to opmath https://github.com/pytorch/pytorch/blob/c3e67d8a8c1217c6c65c0539866e6c59ee2bd9a4/aten/src/ATen/native/cuda/UnarySpecialOpsKernel.cu#L154, whereas sigmoid ref does (all unary refs get wrapped in type promotion wrapper by default). glu ref also needs type promotion wrapper.

Comment (Lezcano) [reply test/test_decomp.py:163]: We looked at it the other day and I completely missed that one! Thank you @ngimel!

Comment (fdrocha) [reply test/test_decomp.py]: Thansk guys. @ngimel I think this should be ready to go, could you PTAL?

Comment (fdrocha) [reply test/test_decomp.py]: Discussed it with @Lezcano and I will also fix up cpu and cuda implementations of sigmoid to do the type promotion for {,b}float16

**Comment by fdrocha** [comment]
Ok, now I think it really is ready. I added the type promotion logic to the cuda kernel. While doing it I noticed that sinc kernel was also missing it so I added it there too. @ngimel and @Lezcano could you PTAL?

**Review by ngimel — APPROVED** [review]
Awesome, looks good!
